### ISSUE TITLE
Make dip switch or service mode available for all FBA games

### ISSUE THREAD

**Issue by bbhenry** [recalbox/recalbox-os#1022]
Many of the CPS1 and CPS2 games supports 3 to 4 players but they are often default to allow only 2 players. In order to change that setting, you'll need to get into the service mode or dip switch to change it. After digging through lots of pages online and testing it with the latest Recalbox beta, I noticed that the service mode is integrated in the Retroarch menu for certain games but not all. Can we have this made available to all the CPS games?

**Comment by Laddyacris** [comment]
Hello
dip switches are availible in fba libretro core in retroarch.
You can create a game-option option file for each CPS1 and CPS2 games

```
Save Configuration On exit > ON
Menu settings > configuration > use per-game core option > ON
Quick Menu > Options
```
Files created storaged : `/recalbox/share/system/configs/retroarch/FB Alpha/namegame.opt`

**Comment by bbhenry** [comment]
@Laddyacris Hi, I was able to follow your instruction in the menus but I can't find the `FB Alpha` directory under `/recalbox/share/system/configs/retroarch/`

Everything under that directory looks like this:

```
# ls -lh
total 64
-rw-r--r--    1 root     root           0 Aug 17 04:39 content_history.lpl
drwxr-xr-x    2 root     root        4.0K Aug 17 04:42 cores
drwxr-xr-x    2 root     root        4.0K Aug 16 05:05 inputs
-rw-r--r--    1 root     root       49.8K Aug 17 04:39 retroarchcustom.cfg
-rw-r--r--    1 root     root        1.2K Jan  1  1970 retroarchcustom.cfg.origin
```

**Comment by bbhenry** [comment]
@Laddyacris 
nevermind, I found it, but the file created for **Battle Circuit** is a empty file. 

`/recalbox/share/system/.config/retroarch/config/FB\ Alpha/batcir.opt` <-- this is a empty file. Isn't it supposed to be created with some options already?

**Comment by Laddyacris** [comment]
Very strange but, i tested few months. The difference are my test on recalbox with retroarch 1.3.1

Look in `/recalbox/system/configs/retroarch/core/retroarch-core-options.cfg`

I look again, tomorrow with retroarch 1.3.4 if different now

in my namegame.opt file there are theses lines :+1: 
`fba-dipswitch-1943kai-Difficulty = "(Default) 8 (Normal)"
fba-dipswitch-1943kai-2_Player_Game = "(Default) 2 Credits / 2 Players"
fba-dipswitch-1943kai-Flip_Screen = "(Default) Off"
fba-dipswitch-1943kai-Freeze = "(Default) Off"
fba-dipswitch-1943kai-Service_Mode = "(Default) Off"
fba-dipswitch-1943kai-Coin_A = "(Default) 1 Coin 1 Play"
fba-dipswitch-1943kai-Coin_B = "(Default) 1 Coin 1 Play"
fba-dipswitch-1943kai-Allow_Continue = "(Default) Yes"
fba-dipswitch-1943kai-Demo_Sounds = "(Default) On"`

**Comment by Laddyacris** [comment]
no service mode for battle circuit : http://www.mamedb.com/game/batcir
If you try Capcom 1941 -> http://www.mamedb.com/game/1941 (dip switches are available but no service mode) but capcom 1943 the battle of midway service mode is available)
```
# ls '/recalbox/share/system/configs/retroarch/FB Alpha/'
1943.opt      aerofgt.opt   batcir.opt    gunbird.opt
1945kiii.opt  afighter.opt  garou.opt     sonicwi2.opt
# ls '/recalbox/share/system/configs/retroarch/FB Alpha/'
1941.opt      1945kiii.opt  afighter.opt  garou.opt     sonicwi2.opt
1943.opt      aerofgt.opt   batcir.opt    gunbird.opt
````

**Comment by bbhenry** [comment]
So is FBA based on Mame 0.147? 

Well, many of the CPS2 games support 3 or 4 players, I haven't tested all of them. But I currently need it for Battle Circuit, and the only way to change it is through service mode or dip switch.

**Comment by Laddyacris** [comment]
It uses the FBA romset version : FBA 0.2.97.37 which is based on MAME 0.167(November 2015)
https://github.com/recalbox/recalbox-os/wiki/Advanced-Arcade-on-Recalbox-%28EN%29

**Comment by bbhenry** [comment]
ok, I was able to use the Mame2003 emulator in Recalbox to run **Battle Circuit** and launch the **service mode** which gave me the capability to enable 4 players as shown in the attatchment. Mame2003 is even older then Mame 0.147(2007). So there is definitely ways to add it to FBA. 

![image](https://cloud.githubusercontent.com/assets/667944/17764545/f0e902ca-64d4-11e6-8fa1-0269355da195.png)
![image](https://cloud.githubusercontent.com/assets/667944/17764594/2a8fb2da-64d5-11e6-9c90-6d592885388a.png)

**Comment by Laddyacris** [comment]
You need to ask libretro team because with fba libretro 0.37 it s not
possible.
Fba libretro core will be update in recalbox 4.1

**Comment by Laddyacris** [comment]
Try to L3 button.

**Comment by bbhenry** [comment]
Is that the only way? because I don't have a joypad that has L3 button

**Comment by dankcushions** [comment]
latest libretro fba has a core option where you can select what button or combo opens the service menu

**Comment by bbhenry** [comment]
@dankcushions cool, but is Recalbox going to use the latest libretro fba in the next release?

**Comment by Laddyacris** [comment]
fba libretro 0.2.97.38 will be add in recalbox 4.1.0

**Comment by rockaddicted** [comment]
> fba libretro 0.2.97.38 will be update in recalbox 4.1.0

I hope, it seems there is some regressions on psykio games with this core version :confused: 
@dankcushions do you have an opinion about that ? Is it an upstream FBA regression ?
http://libretro.com/forums/showthread.php?t=6555

**Comment by Laddyacris** [comment]
Yes, it is a regressoon but in 0.37 psykio game not work very well too.
(Slow ans audio bug)
To play with theses games need fba libretro 0.34.
